File: Assistant/lib/utility/exception_helper.dart
import '/common.dart';
import 'package:flutter/widgets.dart';
import 'package:stack_trace/stack_trace.dart' as lib;

// ----------------

class ExceptionHelper {

  // #region utility

  static Boolean _initialized = false;

  static Void Function(Object exception, StackTrace? stack)? _handler = null;

  // ----------------

  static Void initialize(
  ) {
    assertTest(!ExceptionHelper._initialized);
    WidgetsBinding.instance.platformDispatcher.onError = (exception, stack) {
      ExceptionHelper._handler?.call(exception, stack);
      return true;
    };
    FlutterError.onError = (details) {
      FlutterError.presentError(details);
      ExceptionHelper._handler?.call(details.exception, details.stack);
      return;
    };
    ExceptionHelper._initialized = true;
    return;
  }

  static Void listen(
    Void Function(Object exception, StackTrace? stack)? handler,
  ) {
    assertTest(ExceptionHelper._initialized);
    ExceptionHelper._handler = handler;
    return;
  }

  // ----------------

  static String generateMessage(
    Object      exception,
    StackTrace? stack,
  ) {
    var result = '${exception}';
    if (stack != null) {
      var trace = lib.Trace.from(stack);
      for (var frame in trace.frames) {
        result += '\n@ ${frame.library.selfLet((it) => it.startsWith('package:assistant') ? it.substring('package:'.length) : it)}:${frame.line ?? '?'}:${frame.column ?? '?'} ${frame.member ?? '?'}';
      }
    }
    return result;
  }

  // #endregion

}

File: AssistantPlus/Application/Utility/ExceptionHelper.cs
#pragma warning disable 0,
// ReSharper disable

using Twinning.AssistantPlus;

namespace Twinning.AssistantPlus.Utility {

	public static class ExceptionHelper {

		#region utility

		private static Boolean Initialized { get; set; } = false;

		private static Action<Exception>? Handler { get; set; } = null;

		// ----------------

		public static void Initialize (
			App application
		) {
			AssertTest(!ExceptionHelper.Initialized);
			application.UnhandledException += async (_, args) => {
				args.Handled = true;
				ExceptionHelper.Handler?.Invoke(args.Exception);
				return;
			};
			TaskScheduler.UnobservedTaskException += async (_, args) => {
				args.SetObserved();
				ExceptionHelper.Handler?.Invoke(args.Exception);
				return;
			};
			ExceptionHelper.Initialized = true;
			return;
		}

		public static void Listen (
			Action<Exception>? handler
		) {
			AssertTest(ExceptionHelper.Initialized);
			ExceptionHelper.Handler = handler;
			return;
		}

		// ----------------

		public static Task WrapTask (
			Task task
		) {
			return task.ContinueWith((it) => {
				if (it.Exception?.InnerException != null) {
					ExceptionHelper.Handler?.Invoke(it.Exception.InnerException);
				}
				return;
			});
		}

		public static String GenerateMessage (
			Exception exception
		) {
			var message = $"{exception.Message}";
			var stack = new StackTrace(exception);
			if (exception.StackTrace != null) {
				foreach (var frame in exception.StackTrace.Split(Environment.NewLine)) {
					if (!frame.StartsWith("   at ")) {
						continue;
					}
					message += $"\n@ {frame.Substring("   at ".Length)}";
				}
			}
			return message;
		}

		#endregion

	}

}
